include/RAJA/pattern/detail/reduce.hpp

```diff
@@ -0,0 +1,272 @@
+class BaseReduce
+{
+  using Reduce = Reduce_<T>;
+  // NOTE: the _t here is to appease MSVC
+  using Combiner_t = Combiner_<T, Reduce>;
+  Combiner_t mutable c;
+
+public:
+  using value_type = T;
+  using reduce_type = Reduce;
+
+  //! prohibit compiler-generated default ctor
+  BaseReduce() = delete;
+
+  //! prohibit compiler-generated copy assignment
+  BaseReduce &operator=(const BaseReduce &) = delete;
+
+  //! compiler-generated copy constructor
+  BaseReduce(const BaseReduce &) = default;
+
+  //! compiler-generated move constructor
+  BaseReduce(BaseReduce &&) = default;
+
+  //! compiler-generated move assignment
+  BaseReduce &operator=(BaseReduce &&) = default;
+
+  constexpr BaseReduce(T init_val, T identity_ = Reduce::identity())
+      : c{init_val, identity_}
+  {
+  }
+
+  const BaseReduce &combine(T const &other) const
+  {
+    c.combine(other);
+    return *this;
+  }
+
+  //! Get the calculated reduced value
+  operator T() const { return c.get(); }
+
+  //! Get the calculated reduced value
+  T get() const { return c.get(); }
+};
+
+
+/*!
+ ******************************************************************************
+ *
+ * \brief  Min reducer class template for use in OpenMP execution.
+ *
+ *         For usage example, see reducers.hxx.
+ *
+ ******************************************************************************
+ */
+template <typename T, template <typename, typename> class Combiner>
+class BaseReduceMin : public BaseReduce<T, RAJA::reduce::min, Combiner>
+{
+public:
+  using Base = BaseReduce<T, RAJA::reduce::min, Combiner>;
+  using Base::Base;
+
+  //! reducer function; updates the current instance's state
+  const BaseReduceMin &min(T rhs) const
+  {
+    this->combine(rhs);
+    return *this;
+  }
+};
+
+/*!
+ **************************************************************************
+ *
+ * \brief  MinLoc reducer class template for use in tbb execution.
+ *
+ **************************************************************************
+ */
+template <typename T, template <typename, typename> class Combiner>
+class BaseReduceMinLoc
+    : public BaseReduce<ValueLoc<T>, RAJA::reduce::min, Combiner>
+{
+public:
+  using Base = BaseReduce<ValueLoc<T>, RAJA::reduce::min, Combiner>;
+  using value_type = typename Base::value_type;
+  using Base::Base;
+
+  //! constructor requires a default value for the reducer
+  explicit BaseReduceMinLoc(T init_val, Index_type init_idx)
+      : Base(value_type(init_val, init_idx))
+  {
+  }
+
+  //! constructor requires a default value for the reducer
+  explicit BaseReduceMinLoc(value_type const & init_)
+      : Base(init_)
+  {
+  }
+
+  /// \brief reducer function; updates the current instance's state
+  const BaseReduceMinLoc &minloc(T rhs, Index_type loc) const
+  {
+    this->combine(value_type(rhs, loc));
+    return *this;
+  }
+
+  //! Get the calculated reduced value
+  Index_type getLoc() { return Base::get().getLoc(); }
```

Comment:
shouldn't this be `const` ?

Also, I see a potential to have getLoc() be CRTP inherited -- not a requirement but it does de-duplicate code.

Reply:
Yup, definitely should, thanks for catching that. I'll look at CRTP inheriting some of that too.